js/binary/utils.js

```diff
@@ -1038,7 +1038,7 @@ jspb.utils.stringToByteArray = function(str) {
  * @suppress {invalidCasts}
  */
 jspb.utils.byteSourceToUint8Array = function(data) {
-  if (data.constructor === Uint8Array) {
+  if (data.constructor === Uint8Array || data instanceof Uint8Array) {
```

Comment:
thanks!

instanceof operators are somewhat substantially more expensive than these direct constructor tests.  so it would be best to only use this if we have to

also, this is a bit redundant with the Buffer test on line 1058, in fact i think Buffer will match here now instead of line 1050

finally, due to Buffer now matching up here we have a slight behavior change

I think the best solution would be to add something like

```
 if (data instanceof Uint8Array) {
    // Support types like nodejs Buffer (a subclass of Uint8Array).
    data = /** @type {!Uint8Array} */ (data);
    // Make a shallow copy to ensure we only ever deals with Uint8Array
    // exactly to ensure monomorphism.
    return /** @type {!Uint8Array} */ (
        new Uint8Array(data.buffer, data.byteOffset, data.byteLength));
  }
```

at the end of the method, so we only execute instanceof if we were going to otherwise fail

Reply:
@lukesandberg thanks a lot for the review and your comments. Makes a lot of sense.

I will update the PR and push it up again.

I'm excited to see that this fix is being considered!

Reply:
one thing im not sure i made clear, Given this change, i think we should drop the special case for Buffer and let it be handled by the instanceof test.

Curious what you think about that?